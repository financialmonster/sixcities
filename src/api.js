import axios from 'axios';

const api = axios.create({
	baseURL: `https://htmlacademy-react-3.appspot.com/six-cities`,
    timeout: 5000,
    withCredentials: true
});

const getOffers = () => api.get(`/hotels`);

const sendUserData = (userData) => api.post(`/login`, userData);

const getComments = (id) => api.get(`/comments/${id}`);

const sendUserReview = (review, id) => api.post(`/comments/${id}`, review);

const getFavorites = () => api.get(`/favorite`);

const setFavorite = (id, isFavorite) => api.post(`/favorite/${id}/${(isFavorite) ? 0 : 1}`);

export {getOffers, sendUserData, getComments, sendUserReview, getFavorites, setFavorite};
