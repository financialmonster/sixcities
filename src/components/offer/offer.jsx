import React from 'react';
import PropTypes from 'prop-types';

import Reviews from '../reviews/reviews.jsx';
import ReviewsMap from '../reviewsMap/reviewsMap.jsx';

const Offer = ({offer:{images, is_premium: isPremium, title, is_favorite: isFavorite, rating, max_adults: maxAdults,
    type, bedrooms, price, goods, id, host, description}}) => {
    const className = (host.is_pro)
        ? `property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper`
        : `property__avatar-wrapper user__avatar-wrapper`;

    return (
        <section className="property">
            <div className="property__gallery-container container">
                <div className="property__gallery">
                    {
                        images.slice(0, 6)
                            .map((src) => {
                            return (
                                <div key={src} className="property__image-wrapper">
                                    <img className="property__image" src={src} alt="Photo studio" />
                                </div>
                            );
                        })
                    }
                </div>
            </div>
            <div className="property__container container">
                <div className="property__wrapper">
                    {(isPremium) ? <div className="property__mark"><span>Premium</span></div> : null}

                    <div className="property__name-wrapper">
                        <h1 className="property__name">
                            {title}
                        </h1>
                        <button className="property__bookmark-button button" type="button">
                            <svg className="property__bookmark-icon" width="31" height="33">
                                <use xlinkHref="#icon-bookmark"></use>
                            </svg>
                            <span className="visually-hidden">To bookmarks</span>
                        </button>
                    </div>
                    <div className="property__rating rating">
                        <div className="property__stars rating__stars">
                            <span style={{width: `${20 * rating}%`}}></span>
                            <span className="visually-hidden">Rating</span>
                        </div>
                        <span className="property__rating-value rating__value">{rating}</span>
                    </div>
                    <ul className="property__features">
                        <li className="property__feature property__feature--entire">
                            {type}
                        </li>
                        <li className="property__feature property__feature--bedrooms">
                            {bedrooms} Bedrooms
                        </li>
                        <li className="property__feature property__feature--adults">
                            Max {maxAdults} adults
                        </li>
                    </ul>
                    <div className="property__price">
                        <b className="property__price-value">&euro;{price}</b>
                        <span className="property__price-text">&nbsp;night</span>
                    </div>
                    <div className="property__inside">
                        <h2 className="property__inside-title">What&apos;s inside</h2>
                        <ul className="property__inside-list">
                            {
                                goods.map((item) => {
                                    return (
                                        <li key={item} className="property__inside-item">
                                            {item}
                                        </li>
                                    );
                                })
                            }
                        </ul>
                    </div>
                    <div className="property__host">
                        <h2 className="property__host-title">Meet the host</h2>
                        <div className="property__host-user user">
                            <div className={className}>
                                <img className="property__avatar user__avatar" src={host.avatar_url} width="74"
                                    height="74" alt="Host avatar"
                                />
                            </div>
                            <span className="property__user-name">
                                {host.name}
                            </span>
                        </div>
                        <div className="property__description">
                            <p className="property__text">
                                {description}
                            </p>
                        </div>
                    </div>
                    
                    <Reviews id={id} />
                </div>
            </div>
            <ReviewsMap />
        </section>
    );
}

Offer.propTypes = {
    offer: PropTypes.shape({
        price: PropTypes.number.isRequired,
        is_favorite: PropTypes.bool.isRequired,
        is_premium: PropTypes.bool.isRequired,
        rating: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired,
        type: PropTypes.oneOf([`room`, `house`, `hotel`, `apartment`]).isRequired,
        preview_image: PropTypes.string.isRequired,
        id: PropTypes.number.isRequired,
        images: PropTypes.arrayOf(PropTypes.string).isRequired,
        bedrooms: PropTypes.number.isRequired,
        max_adults: PropTypes.number.isRequired,
        goods: PropTypes.arrayOf(PropTypes.string).isRequired,
        host: PropTypes.shape({
            id: PropTypes.number.isRequired,
            is_pro: PropTypes.bool.isRequired,
            name: PropTypes.string.isRequired,
            avatar_url: PropTypes.string.isRequired
        }),
        description: PropTypes.string.isRequired,
        location: PropTypes.shape({
            latitude: PropTypes.number.isRequired,
            longitude: PropTypes.number.isRequired,
            zoom: PropTypes.number.isRequired
        })
    })
}

export {Offer};
export default Offer;
