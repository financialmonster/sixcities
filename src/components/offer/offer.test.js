import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import {BrowserRouter as Router} from 'react-router-dom';
import thunk from 'redux-thunk';
import api from '../../api.js';

import {Offer} from './offer.jsx';

const favoriteOffer = {preview_image: "https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg", id: 2,
    images: ["https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"], rating: 3.4, type: `hotel`,
    title:"Penthouse, 4-5 rooms + 5 balconies", is_favorite: true, is_premium: false,  bedrooms: 2, max_adults: 8,
    price: 471, goods: ["Baby seat"], description: "Peaceful studio in the most wanted area in town.",
    host: {id: 25, name: "Angelina", is_pro: true, avatar_url: "img/avatar-angelina.jpg"},
    location: {latitude: 50.913361, longitude: 6.9509739999999995, zoom: 16}
};

const commonOffer = {preview_image: "https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg", id: 1,
    images: ["https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"], rating: 3.4, type: `hotel`,
    title:"Penthouse, 4-5 rooms + 5 balconies", is_favorite: false, is_premium: false,  bedrooms: 2, max_adults: 8,
    price: 471, goods: ["Baby seat"], description: "Peaceful studio in the most wanted area in town.",
    host: {id: 25, name: "Angelina", is_pro: true, avatar_url: "img/avatar-angelina.jpg"},
    location: {latitude: 50.913361, longitude: 6.9509739999999995, zoom: 16}
};

describe(`Correctly renders the Offer component`, () => {
    it(`when isFavoriteLoading property is equal true`, () => {
        const mockStore = configureStore([thunk.withExtraArgument(api)]);
        const store = mockStore({
            card: {focusedCard: null},
            nearby: {nearbyHotels: []},
            offers: {isTestMode: true},
            auth: {isLoggedIn: false, isAuthStatusLoading: false , isAuthStatusError: false},
            reviews: {isReviewsError: false, isReviewsLoading: false , reviews: [], isReviewError: false,
                isReviewLoading: false}
        });

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <Offer fetchFavorite={() => {}} isFavoriteLoading={true} isFavoriteError={false} isLoggedIn={true}
                        offer={commonOffer} history={{}}
                    />
                </Router>
            </Provider>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isFavoriteError property is equal true`, () => {
        const mockStore = configureStore([thunk.withExtraArgument(api)]);
        const store = mockStore({
            card: {focusedCard: null},
            nearby: {nearbyHotels: []},
            offers: {isTestMode: true},
            auth: {isLoggedIn: false, isAuthStatusLoading: false , isAuthStatusError: false},
            reviews: {isReviewsError: false, isReviewsLoading: false , reviews: [], isReviewError: false,
                isReviewLoading: false
            }
        });

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <Offer fetchFavorite={() => {}} isFavoriteLoading={false} isFavoriteError={true} isLoggedIn={true}
                        offer={commonOffer} history={{}} 
                    />
                </Router>
            </Provider>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isFavoriteError, is_favorite and isFavoriteLoading properties are equal false`, () => {
        const mockStore = configureStore([thunk.withExtraArgument(api)]);
        const store = mockStore({
            card: {focusedCard: null},
            nearby: {nearbyHotels: []},
            offers: {isTestMode: true},
            auth: {isLoggedIn: false, isAuthStatusLoading: false , isAuthStatusError: false},
            reviews: {isReviewsError: false, isReviewsLoading: false , reviews: [], isReviewError: false,
                isReviewLoading: false
            }
        });

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <Offer fetchFavorite={() => {}} isFavoriteLoading={false} isFavoriteError={false} isLoggedIn={true}
                        offer={commonOffer} history={{}} 
                    />
                </Router>
            </Provider>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`when isFavoriteError and isFavoriteLoading properties are equal false, but is_favorite property is equal
        true`,
    () => {
        const mockStore = configureStore([thunk.withExtraArgument(api)]);
        const store = mockStore({
            card: {focusedCard: null},
            nearby: {nearbyHotels: []},
            offers: {isTestMode: true},
            auth: {isLoggedIn: false, isAuthStatusLoading: false , isAuthStatusError: false},
            reviews: {isReviewsError: false, isReviewsLoading: false , reviews: [], isReviewError: false,
                isReviewLoading: false
            }
        });

        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <Offer fetchFavorite={() => {}} isFavoriteLoading={false} isFavoriteError={false} isLoggedIn={true}
                        offer={favoriteOffer} history={{}} 
                    />
                </Router>
            </Provider>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
