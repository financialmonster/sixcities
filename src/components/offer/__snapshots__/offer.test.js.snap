// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly renders the Offer component when isFavoriteError and isFavoriteLoading properties are equal false, but is_favorite property is equal
        true 1`] = `
<section
  className="property"
>
  <div
    className="property__gallery-container container"
  >
    <div
      className="property__gallery"
    >
      <div
        className="property__image-wrapper"
      >
        <img
          alt="Studio"
          className="property__image"
          src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"
        />
      </div>
    </div>
  </div>
  <div
    className="property__container container"
  >
    <div
      className="property__wrapper"
    >
      <div
        className="property__name-wrapper"
      >
        <h1
          className="property__name"
        >
          Penthouse, 4-5 rooms + 5 balconies
        </h1>
        <button
          className="property__bookmark-button button property__bookmark-button--active"
          onClick={[Function]}
          title="Remove from bookmarks"
          type="button"
        >
          <svg
            className="property__bookmark-icon"
            height="33"
            width="31"
          >
            <use
              xlinkHref="#icon-bookmark-active"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            In bookmarks
          </span>
        </button>
      </div>
      <div
        className="property__rating rating"
      >
        <div
          className="property__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "68%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
        <span
          className="property__rating-value rating__value"
        >
          3.4
        </span>
      </div>
      <ul
        className="property__features"
      >
        <li
          className="property__feature property__feature--entire"
        >
          hotel
        </li>
        <li
          className="property__feature property__feature--bedrooms"
        >
          2
           Bedrooms
        </li>
        <li
          className="property__feature property__feature--adults"
        >
          Max 
          8
           adults
        </li>
      </ul>
      <div
        className="property__price"
      >
        <b
          className="property__price-value"
        >
          €
          471
        </b>
        <span
          className="property__price-text"
        >
           night
        </span>
      </div>
      <div
        className="property__inside"
      >
        <h2
          className="property__inside-title"
        >
          What's inside
        </h2>
        <ul
          className="property__inside-list"
        >
          <li
            className="property__inside-item"
          >
            Baby seat
          </li>
        </ul>
      </div>
      <div
        className="property__host"
      >
        <h2
          className="property__host-title"
        >
          Meet the host
        </h2>
        <div
          className="property__host-user user"
        >
          <div
            className="property__avatar-wrapper user__avatar-wrapper property__avatar-wrapper--pro"
          >
            <img
              alt="Host avatar"
              className="property__avatar user__avatar"
              height="74"
              src="avatar-angelina.jpg"
              width="74"
            />
          </div>
          <span
            className="property__user-name"
          >
            Angelina
          </span>
        </div>
        <div
          className="property__description"
        >
          <p
            className="property__text"
          >
            Peaceful studio in the most wanted area in town.
          </p>
        </div>
      </div>
      <section
        className="property__reviews reviews"
      >
        <h2
          className="reviews__title"
        >
          Reviews · 
          <span
            className="reviews__amount"
          >
            0
          </span>
        </h2>
        <ul
          className="reviews__list"
        />
      </section>
    </div>
  </div>
  <section
    className="property__map map"
  >
    <div
      id="reviewsMap"
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </section>
</section>
`;

exports[`Correctly renders the Offer component when isFavoriteError property is equal true 1`] = `
<section
  className="property"
>
  <div
    className="property__gallery-container container"
  >
    <div
      className="property__gallery"
    >
      <div
        className="property__image-wrapper"
      >
        <img
          alt="Studio"
          className="property__image"
          src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"
        />
      </div>
    </div>
  </div>
  <div
    className="property__container container"
  >
    <div
      className="property__wrapper"
    >
      <div
        className="property__name-wrapper"
      >
        <h1
          className="property__name"
        >
          Penthouse, 4-5 rooms + 5 balconies
        </h1>
        <button
          className="property__bookmark-button button"
          onClick={[Function]}
          title="Add to bookmarks"
          type="button"
        >
          <div>
            Fail
          </div>
        </button>
      </div>
      <div
        className="property__rating rating"
      >
        <div
          className="property__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "68%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
        <span
          className="property__rating-value rating__value"
        >
          3.4
        </span>
      </div>
      <ul
        className="property__features"
      >
        <li
          className="property__feature property__feature--entire"
        >
          hotel
        </li>
        <li
          className="property__feature property__feature--bedrooms"
        >
          2
           Bedrooms
        </li>
        <li
          className="property__feature property__feature--adults"
        >
          Max 
          8
           adults
        </li>
      </ul>
      <div
        className="property__price"
      >
        <b
          className="property__price-value"
        >
          €
          471
        </b>
        <span
          className="property__price-text"
        >
           night
        </span>
      </div>
      <div
        className="property__inside"
      >
        <h2
          className="property__inside-title"
        >
          What's inside
        </h2>
        <ul
          className="property__inside-list"
        >
          <li
            className="property__inside-item"
          >
            Baby seat
          </li>
        </ul>
      </div>
      <div
        className="property__host"
      >
        <h2
          className="property__host-title"
        >
          Meet the host
        </h2>
        <div
          className="property__host-user user"
        >
          <div
            className="property__avatar-wrapper user__avatar-wrapper property__avatar-wrapper--pro"
          >
            <img
              alt="Host avatar"
              className="property__avatar user__avatar"
              height="74"
              src="avatar-angelina.jpg"
              width="74"
            />
          </div>
          <span
            className="property__user-name"
          >
            Angelina
          </span>
        </div>
        <div
          className="property__description"
        >
          <p
            className="property__text"
          >
            Peaceful studio in the most wanted area in town.
          </p>
        </div>
      </div>
      <section
        className="property__reviews reviews"
      >
        <h2
          className="reviews__title"
        >
          Reviews · 
          <span
            className="reviews__amount"
          >
            0
          </span>
        </h2>
        <ul
          className="reviews__list"
        />
      </section>
    </div>
  </div>
  <section
    className="property__map map"
  >
    <div
      id="reviewsMap"
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </section>
</section>
`;

exports[`Correctly renders the Offer component when isFavoriteError, is_favorite and isFavoriteLoading properties are equal false 1`] = `
<section
  className="property"
>
  <div
    className="property__gallery-container container"
  >
    <div
      className="property__gallery"
    >
      <div
        className="property__image-wrapper"
      >
        <img
          alt="Studio"
          className="property__image"
          src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"
        />
      </div>
    </div>
  </div>
  <div
    className="property__container container"
  >
    <div
      className="property__wrapper"
    >
      <div
        className="property__name-wrapper"
      >
        <h1
          className="property__name"
        >
          Penthouse, 4-5 rooms + 5 balconies
        </h1>
        <button
          className="property__bookmark-button button"
          onClick={[Function]}
          title="Add to bookmarks"
          type="button"
        >
          <svg
            className="property__bookmark-icon"
            height="33"
            width="31"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            To bookmarks
          </span>
        </button>
      </div>
      <div
        className="property__rating rating"
      >
        <div
          className="property__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "68%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
        <span
          className="property__rating-value rating__value"
        >
          3.4
        </span>
      </div>
      <ul
        className="property__features"
      >
        <li
          className="property__feature property__feature--entire"
        >
          hotel
        </li>
        <li
          className="property__feature property__feature--bedrooms"
        >
          2
           Bedrooms
        </li>
        <li
          className="property__feature property__feature--adults"
        >
          Max 
          8
           adults
        </li>
      </ul>
      <div
        className="property__price"
      >
        <b
          className="property__price-value"
        >
          €
          471
        </b>
        <span
          className="property__price-text"
        >
           night
        </span>
      </div>
      <div
        className="property__inside"
      >
        <h2
          className="property__inside-title"
        >
          What's inside
        </h2>
        <ul
          className="property__inside-list"
        >
          <li
            className="property__inside-item"
          >
            Baby seat
          </li>
        </ul>
      </div>
      <div
        className="property__host"
      >
        <h2
          className="property__host-title"
        >
          Meet the host
        </h2>
        <div
          className="property__host-user user"
        >
          <div
            className="property__avatar-wrapper user__avatar-wrapper property__avatar-wrapper--pro"
          >
            <img
              alt="Host avatar"
              className="property__avatar user__avatar"
              height="74"
              src="avatar-angelina.jpg"
              width="74"
            />
          </div>
          <span
            className="property__user-name"
          >
            Angelina
          </span>
        </div>
        <div
          className="property__description"
        >
          <p
            className="property__text"
          >
            Peaceful studio in the most wanted area in town.
          </p>
        </div>
      </div>
      <section
        className="property__reviews reviews"
      >
        <h2
          className="reviews__title"
        >
          Reviews · 
          <span
            className="reviews__amount"
          >
            0
          </span>
        </h2>
        <ul
          className="reviews__list"
        />
      </section>
    </div>
  </div>
  <section
    className="property__map map"
  >
    <div
      id="reviewsMap"
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </section>
</section>
`;

exports[`Correctly renders the Offer component when isFavoriteLoading property is equal true 1`] = `
<section
  className="property"
>
  <div
    className="property__gallery-container container"
  >
    <div
      className="property__gallery"
    >
      <div
        className="property__image-wrapper"
      >
        <img
          alt="Studio"
          className="property__image"
          src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"
        />
      </div>
    </div>
  </div>
  <div
    className="property__container container"
  >
    <div
      className="property__wrapper"
    >
      <div
        className="property__name-wrapper"
      >
        <h1
          className="property__name"
        >
          Penthouse, 4-5 rooms + 5 balconies
        </h1>
        <button
          className="property__bookmark-button button"
          onClick={[Function]}
          title="Add to bookmarks"
          type="button"
        >
          <div>
            ...
          </div>
        </button>
      </div>
      <div
        className="property__rating rating"
      >
        <div
          className="property__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "68%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
        <span
          className="property__rating-value rating__value"
        >
          3.4
        </span>
      </div>
      <ul
        className="property__features"
      >
        <li
          className="property__feature property__feature--entire"
        >
          hotel
        </li>
        <li
          className="property__feature property__feature--bedrooms"
        >
          2
           Bedrooms
        </li>
        <li
          className="property__feature property__feature--adults"
        >
          Max 
          8
           adults
        </li>
      </ul>
      <div
        className="property__price"
      >
        <b
          className="property__price-value"
        >
          €
          471
        </b>
        <span
          className="property__price-text"
        >
           night
        </span>
      </div>
      <div
        className="property__inside"
      >
        <h2
          className="property__inside-title"
        >
          What's inside
        </h2>
        <ul
          className="property__inside-list"
        >
          <li
            className="property__inside-item"
          >
            Baby seat
          </li>
        </ul>
      </div>
      <div
        className="property__host"
      >
        <h2
          className="property__host-title"
        >
          Meet the host
        </h2>
        <div
          className="property__host-user user"
        >
          <div
            className="property__avatar-wrapper user__avatar-wrapper property__avatar-wrapper--pro"
          >
            <img
              alt="Host avatar"
              className="property__avatar user__avatar"
              height="74"
              src="avatar-angelina.jpg"
              width="74"
            />
          </div>
          <span
            className="property__user-name"
          >
            Angelina
          </span>
        </div>
        <div
          className="property__description"
        >
          <p
            className="property__text"
          >
            Peaceful studio in the most wanted area in town.
          </p>
        </div>
      </div>
      <section
        className="property__reviews reviews"
      >
        <h2
          className="reviews__title"
        >
          Reviews · 
          <span
            className="reviews__amount"
          >
            0
          </span>
        </h2>
        <ul
          className="reviews__list"
        />
      </section>
    </div>
  </div>
  <section
    className="property__map map"
  >
    <div
      id="reviewsMap"
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </section>
</section>
`;
